### PR TITLE
Deprecated `Trainer.log_dir` with `Trainer.experiment_dir`

### LINKED ISSUE #14188: Introduce `Logger.experiment_dir`
## 🚀 Feature

Introduce a property/method `experiment_dir` on the abstract logger interface. All loggers must implement this and return a path to where their files are being saved. 

### Motivation

The Trainer currently determines where checkpoints are saved like so:
```py
os.path.join(logger.save_dir, str(logger.name), logger.version, "checkpoints")
```

As you can see, it assumes that there is a rigid folder structure `save_dir/name/version`. Every logger has to follow this structure.
However, at least one logger (wandb) struggles with this, because it has an additional concept of "project" that does not fit well within this layout (https://github.com/Lightning-AI/lightning/issues/14054). Furthermore, the "name" may change AFTER the files were written, so there is a disconnect between the folder structure and what is represented "online" in the wandb ui. 

One observation is that the trainer actually never cares about any of the subfolders. All it needs is the full path. This is why I believe, if we let the logger decide how to assemble the path, we will avoid many issues and disagreements with 3rd party experiment management. 

We are struggling with making all loggers agree to a standardized layout, so maybe we shouldn't try to do that, at least not in every aspect of how the loggers work (https://github.com/Lightning-AI/lightning/issues/12028#issuecomment-1088325894). 

### Pitch

Introduce

```py
def experiment_dir(self) -> str:
    """Returns the (relative) path to where logs and artifacts get saved for the current version."""

```

This should probably be a relative path, so that the Trainer can prepend a root dir if needed (default cwd). 
An implementation of a logger could then look like this:

```py
def experiment_dir(self):
    return os.path.join(self.save_dir, self.name, self.version)
```

And the checkpointing in the Trainer would save files to

```py
os.path.join(logger.experiment_dir(), "checkpoints") 
```
for example.

### Implications

1. After introducing experiment dir, we could consider deprecating `save_dir` and `log_dir` or remove it from the base interface.
2. The code here in `Trainer.log_dir` would probably get simplified: https://github.com/Lightning-AI/lightning/blob/fe9e5d55bf7991ba36b76d6adae9075b93dfcaa0/src/pytorch_lightning/trainer/trainer.py#L2208-L2219
3. The code in the checkpoint callback would get simplified too: https://github.com/Lightning-AI/lightning/blob/fe9e5d55bf7991ba36b76d6adae9075b93dfcaa0/src/pytorch_lightning/callbacks/model_checkpoint.py#L595-L608
4. It would immediately address these issues, as the LightningCLI can write to the experiment dir without ambiguity: #14162, #12748

### Alternatives

Struggle

### Additional context

Discussions in:
https://github.com/Lightning-AI/lightning/issues/14054

Current blockers: 
#12177
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @tchaton @justusschock @awaelchli @edward-io @ananthsub @rohitgr7 @kamil-kaczmarek @Raalsky @Blaizzy

### PR DESCRIPTION
**Pull request by tshu-w**
## What does this PR do?

Fixes #14188
Includes and closes #14508

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, make `Trainer.log_dir` to use `Logger.experiment_dir` and ModelCheckpoint use `Trainer.log_dir`  directly.

Next steps:
2. Update implementations of some Loggers, and maybe drop some unnecessary properties(`name`, `version`, etc)
3. Update trainer doc

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
@tshu-w CI is red, have you checked the failing tests?

File: src/pytorch_lightning/callbacks/model_checkpoint.py [hunk 1]
```diff
@@ -591,20 +591,9 @@ def __resolve_ckpt_dir(self, trainer: "pl.Trainer") -> None:
...
-            name = trainer.loggers[0].name
-            version = trainer.loggers[0].version
-            version = version if isinstance(version, str) else f"version_{version}"
-
-            ckpt_path = os.path.join(save_dir, str(name), version, "checkpoints")
         else:
-            # if no loggers, use default_root_dir
-            ckpt_path = os.path.join(trainer.default_root_dir, "checkpoints")
+            log_dir = trainer.log_dir if trainer.log_dir is not None else trainer.default_root_dir
+            ckpt_path = os.path.join(log_dir, "checkpoints")
```
Comment (awaelchli) [inline src/pytorch_lightning/callbacks/model_checkpoint.py:596]: We might want to introduce `Trainer.experiment_dir` for this purpose here, then deprecate the old `Trainer.log_dir`. For users who access trainer.log_dir in their code. 

What do you think?

**Comment by tshu-w** [comment]
> @tshu-w CI is red, have you checked the failing tests?

I have fixed most of them ;)

I don't know how to fix the last one, if I add `print(logger.experiment)` between `fit` and `test`, the test pass on my side. https://github.com/Lightning-AI/lightning/blob/024e7b8204846ead790cf5a098e4a1ca61fcb171/tests/tests_pytorch/loggers/test_all.py#L119-L141

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: Good idea, but I find that this will bring many modifications, should I start another PR or follow this one?

Comment (awaelchli) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: Do you mean because trainer.log_dir is used in many places? I think that's fine. If you think it's too much, you could only do the changes in the source code and do the docs updates in a separate PR mabye.

Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: I see, will look at it. BTW, it will be much appreciated if you could check the last failed test (`loggers/test_all.py::test_loggers_fit_test_all[CometLogger]`) when you are free, I'm not sure why it failed.

Comment (awaelchli) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: I gave it a try. Not 100% sure if my fix is correct, but if I'm looking at it correctly, it's just because the removed logic in model checkpoint caused the test to first access version before experiment, and because we mock a lot of things it didn't like that :)

File: src/pytorch_lightning/CHANGELOG.md [hunk 2]
```diff
@@ -40,6 +40,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Added `WandbLogger.download_artifact` and `WandbLogger.use_artifact` for managing artifacts with Weights and Biases ([#14551](https://github.com/Lightning-AI/lightning/issues/14551))
 
 
+- Introduce `Logger.experiement_dir` ([#14188](https://github.com/Lightning-AI/lightning/pull/14508))
+
+
```
Comment (Borda) [inline src/pytorch_lightning/CHANGELOG.md:45]:
```suggestion
```

File: src/pytorch_lightning/CHANGELOG.md [hunk 3]
```diff
@@ -72,6 +75,12 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - In Lightning Lite, state-dict access to the module wrapper now gets passed through to the original module reference ([#14629](https://github.com/Lightning-AI/lightning/pull/14629))
 
 
+- Improved the error messaging when passing `Trainer.method(model, x_dataloader=None)` with no module-method implementations available ([#14614](https://github.com/Lightning-AI/lightning/pull/14614))
+
+
```
Comment (Borda) [inline src/pytorch_lightning/CHANGELOG.md:80]:
```suggestion
```

File: src/pytorch_lightning/callbacks/model_checkpoint.py (hunk 1 above)
Comment (tshu-w) [reply src/pytorch_lightning/callbacks/model_checkpoint.py]: Thx @awaelchli, I'm even less sure because I'm not familiar with the Trainer part yet, so I didn't modify the test directly. 

Also I have submitted the code to introduce `Trainer.experiment_dir`.

**Comment by Borda** [comment]
@carmocca @awaelchli, mind checking this one? :otter:

**Comment by tshu-w** [comment]
> @Borda I believe we are quite close to the 1.8 release and my capacity is already used up at the moment. In order to finish #14188 completely, we need to properly sequence all (three?) PRs that @tshu-w started.

I've submitted two PRs so far, and this one follows (or includes) https://github.com/Lightning-AI/lightning/pull/14508, so if there's no problem, we can just merge this one.

I'll check the failing test about docs this weekend.

**Comment by tshu-w** [comment]
This PR is ready to review now.

File: src/pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -1863,18 +1862,27 @@ def model(self, model: torch.nn.Module) -> None:
...
+                save_dir = self.default_root_dir
+            dirpath = os.path.join(save_dir, self.loggers[0].experiment_dir)
         else:
             dirpath = self.default_root_dir
 
         dirpath = self.strategy.broadcast(dirpath)
         return dirpath
 
+    @property
+    def log_dir(self) -> str:
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/trainer.py:1879]: Where did we discuss that `log_dir` would be deprecated?

Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:1879]: If they have different purposes, then I don't think `log_dir` should be deprecated. There would be uses cases for each.

If they have the same purpose and just differ in implementation, then we could just change it without a deprecation and keep the `log_dir` name.
In the past, we've done changes to `log_dir` without going through a deprecation.

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:1879]: @carmocca The reasoning is the following:

I think we should make `log_dir` do what it does for our default logger, TensorBoard, meaning returning the full log dir `root / name / version`. This would mean calling log_dir after this PR is the same for these users, no breaking change.

BUT, the consideration for deprecation is only relevant for everyone else who doesn't use TB. Here, we have the option to 
- breaking change, no warning, just change it
- forward deprecation: warn that the behavior will change in the future. Problem: User can't do anything to avoid the warning
- Deprecate the entire log_dir: Same as above, but user can change to trainer.experiment_dir to avoid the deprecation warning

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:1879]: There is also the cheap alternative to masquerade it as a fix, i.e., pushing the change as "Fixed: Trainer.log_dir is now consistently returning the versioned log dir for all loggers".

Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:1879]: Thanks for the context. I think you should choose.

My only suggestion is that this is hard to understand. We would have `log_dir`, `experiment_dir`, `save_dir`, `default_root_dir` ... (any other dir?) and the relationship is not clear from just the names.
If the solution can avoid introducing a new name, I would vote for it. `log_dir` is for the `Trainer`, but would it make sense to have it also be the name for `Logger.log_dir`?

Comment (tshu-w) [reply src/pytorch_lightning/trainer/trainer.py:1879]: Hi, any update on it?
I vote for deprecate the entire `log_dir` or  keep `log_dir`

**Comment by Borda** [comment]
@carmocca, what is missing here? :rabbit:

**Comment by carmocca** [comment]
We need to decide the resolution to the conversation above

File: src/pytorch_lightning/trainer/trainer.py (hunk 4 above)
Comment (carmocca) [reply src/pytorch_lightning/trainer/trainer.py:1879]: I lean towards the latter. @awaelchli what's your opinion? cc @justusschock too

Comment (awaelchli) [reply src/pytorch_lightning/trainer/trainer.py:1879]: @carmocca It is not just about picking the name. The main challenge is how to deliver the improved way without major breaking changes. We can imagine that logging and managing experiments is a core Lightning experience, so it will affect many users. Introducing experiment dir was meant to minimize that (existing code still works). I would also love to keep the same name log_dir, but how are we going to do it???

Comment (justusschock) [reply src/pytorch_lightning/trainer/trainer.py:1879]: I would also try to keep the naming as unified as possible and go with `Logger.log_dir`.

**Comment by tshu-w** [comment]
Hi there, any updates on the resolution? I noticed that @awaelchli  submitted a pull request at https://github.com/Lightning-AI/lightning/pull/16663, but closed it.

**Comment by Borda** [comment]
ping @awaelchli @justusschock @carmocca ^^

**Comment by lantiga** [comment]
I’m against breaking changes in this case

**Comment by tshu-w** [comment]
I think https://github.com/Lightning-AI/lightning/issues/14188 still needs to be resolved. If breaking changes are not supported, can we consider the other two solutions mentioned in https://github.com/Lightning-AI/lightning/pull/14640#discussion_r981731009?

**Comment by lantiga** [comment]
100% with you there @tshu-w
What solution of the two would you recommend?

**Comment by tshu-w** [comment]
Both are fine for me, it's up to you to discuss and decide.😃
